### ISSUE TITLE
PR 5 — Peak-finding vectorization + backend unification unlock

### ISSUE THREAD

**Issue by gitttt-1234** [talmolab/sleap-nn#513]
**Parent epic:** #508
**Prerequisites:** #509
**Blocks:** #515

---

## Goal

Replace the Python-loop-over-peaks integral refinement path with a single vectorized `F.grid_sample` call. This PR is doing **two things at once**:

1. **Performance**: benchmarks show peak finding is 22-62% of total GPU inference time, and integral refinement is 30-50% of that. Target: 2× on `find_local_peaks` integral path.
2. **Unification**: `F.grid_sample` exports cleanly to ONNX (opset ≥ 16), so after this PR the Torch integral-refinement path and a (new) ONNX integral-refinement path can share a single implementation in `ops/peaks.py`. Without this PR, #515 cannot bake integral refinement into ONNX wrappers. So this is **also the unification unlock** — not just performance.

Per design doc §4C: `ops/peaks.py` and `ops/crops.py` become the single source of truth for both the Torch inference path and the ONNX export graph. Current duplicate ops in `export/wrappers/base.py` (`_find_global_peaks`, `_find_topk_peaks`, `_neighbor_max`) are deleted in #515 and replaced with direct calls to `ops/`.

## Why this matters

End-to-end breakdown on **CUDA (NVIDIA A40, torch 2.9.1+cu128, 2026-04-29)** — single-instance UNet, 384×384 grayscale (`benchmarks_cuda/05_end_to_end_results.txt`):

| Batch | Model forward | Peak finding | Peak % of total |
|---:|---:|---:|---:|
| 1 | 0.80 ms | **1.72 ms** | **66.3%** ← peak finding > forward |
| 4 | 1.60 ms | 1.64 ms | 48.6% |
| 16 | 5.89 ms | 1.56 ms | 20.1% |

Almost identical pattern to MPS (62%/42%/22% in the same harness). On any GPU at low batch sizes, peak finding is **bigger than the model forward**. The design plan's prior claim ("already optimized") is about the dilation kernel, not the full path — the remaining cost is in the Python-level integral-refinement patch loop, which is now confirmed dominant on real CUDA hardware.

Original MPS measurements (kept for reference):

| Batch | Model forward | Peak finding | Peak % of total |
|---:|---:|---:|---:|
| 1 | 1.28 ms | 3.26 ms | 62% |
| 4 | 4.23 ms | 3.55 ms | 42% |
| 16 | 16.28 ms | 4.88 ms | 22% |

Per-op breakdown from `benchmarks/01_peak_finding.py` (CPU baseline, with **CUDA confirmation** at the worst case):

| Call | B=1, 192² | B=4, 192² | B=16, 384² (CPU) | B=16, 384² (CUDA A40) |
|---|---:|---:|---:|---:|
| `find_global_peaks_rough` | 0.21 ms | 0.64 ms | 2.53 ms | 0.10 ms |
| `find_global_peaks` (integral) | 0.44 ms | 1.20 ms | 4.49 ms | 1.66 ms |
| `find_local_peaks_rough` | 6.21 ms | 22.90 ms | **407 ms** | **26.80 ms** |
| `find_local_peaks` (integral) | 10.04 ms | 36.40 ms | **626 ms** | **43.20 ms** |

CUDA brings `find_local_peaks` integral down 14× vs CPU at B=16/384², but the **integral-refinement overhead remains 38% on top of rough** — same shape as on CPU, same Python-loop bottleneck. The 2× target is realistic.

## Why `F.grid_sample`

The ONNX bottom-up wrapper already uses this approach — see `sleap_nn/export/wrappers/bottomup.py::_sample_and_score_lines`. It's a proven pattern: `grid_sample` fuses sampling + padding + bilinear-interp into one GPU kernel, no Python loop, fully vectorized.

## Proposed change

Currently (`ops/peaks.py` after #509):

```python
rough_peaks = find_local_peaks_rough(cms)
bboxes = make_centered_bboxes(rough_peaks, patch_size, patch_size)
cm_crops = crop_bboxes(cms_flat, bboxes, box_sample_inds)   # ← slow: unfold + index per peak
dx, dy = integral_regression(cm_crops, gv, gv)
refined = rough_peaks + offsets
```

Proposed:

```python
rough_peaks = find_local_peaks_rough(cms)

# Build a grid of (patch_size × patch_size) offsets around each peak.
# Normalize to [-1, 1] space expected by grid_sample.
grid = build_sample_grid(rough_peaks, patch_size, cms.shape)
# grid shape: (N_peaks, patch_size, patch_size, 2)

# Single GPU kernel extracts all patches.
cm_crops = F.grid_sample(
    cms_flat, grid,
    mode="bilinear", padding_mode="zeros", align_corners=False,
)

dx, dy = integral_regression(cm_crops, gv, gv)   # already vectorized
refined = rough_peaks + offsets
```

### `build_sample_grid` (new helper)

```python
def build_sample_grid(peaks, patch_size, img_shape):
    """Build a (N_peaks, patch_size, patch_size, 2) grid of normalized sampling coords.

    Each grid centers on a peak and extends ±patch_size/2 in each direction.
    Output is in [-1, 1] normalized coordinates per grid_sample convention.
    """
    B, C, H, W = img_shape
    half = (patch_size - 1) / 2

    # Per-peak coordinate meshgrid (patch_size × patch_size)
    dy, dx = torch.meshgrid(
        torch.arange(-half, half + 1, device=peaks.device),
        torch.arange(-half, half + 1, device=peaks.device),
        indexing="ij",
    )
    # Shape: (patch_size, patch_size, 2) — offsets from peak center
    offsets = torch.stack([dx, dy], dim=-1)

    # Broadcast to (N_peaks, patch_size, patch_size, 2)
    peak_xy = peaks.view(-1, 1, 1, 2)
    abs_coords = peak_xy + offsets

    # Normalize to [-1, 1]
    abs_coords[..., 0] = (abs_coords[..., 0] / (W - 1)) * 2 - 1
    abs_coords[..., 1] = (abs_coords[..., 1] / (H - 1)) * 2 - 1
    return abs_coords
```

## Also in this PR: vectorize `crop_bboxes`

`crop_bboxes` is used elsewhere (top-down stage-2 crop extraction) with the same variable-count loop problem. The top-down ONNX wrapper already uses `F.grid_sample` for this (see `export/wrappers/topdown.py::_extract_crops`). Port the same pattern to `ops/crops.py::crop_bboxes`.

## Target numbers

| Metric | Current | Target |
|---|---:|---:|
| `find_local_peaks` integral (B=16, 384², CPU) | 626 ms | ≤ 300 ms (2×) |
| `find_local_peaks` integral (B=16, 384², CUDA A40) | 43.2 ms | ≤ 22 ms (2×) |
| `find_global_peaks` integral (B=1, MPS) | ~3.26 ms | ≤ 2.0 ms (1.6×) |
| `find_global_peaks` integral (B=1, CUDA A40) | 1.66 ms | ≤ 1.0 ms (1.6×) |
| Total GPU inference time (bottom-up) | baseline | −10 to −20% |

Numbers measured by `benchmarks/01_peak_finding.py` — this script must be re-run before and after to document the delta.

## Determinism + parity test plan

The new path (`F.grid_sample` bilinear) is **not bit-exact** with the old path (`unfold` + per-peak index): bilinear weights at integer-pixel centers and zero-padding at borders differ by rounding from the unfold/gather variant. The 1e-4 tolerance accounts for that delta — but the breadth of the test, not the tolerance, is what catches regressions on edge-case inputs.

Required parity surface (broader than the original "compare on one fixture"):

1. **Per-checkpoint golden snapshots.** For each test checkpoint that uses peak finding (single-instance, top-down stage 2, centroid, bottom-up, top-down multi-class, bottom-up multi-class — six total), record the `pred_keypoints` / `pred_centroids` numpy array from the **old** code and check it in as a golden fixture. The new path must reproduce within 1e-4 abs / 1e-5 rel tolerance per checkpoint.

2. **Close-peak stress fixture.** Construct a synthetic confmap with multiple peaks within `<2 * integral_patch_size` of each other (where overlapping bilinear neighborhoods bite hardest — most relevant for bottom-up at high instance density). Old vs new must match within 1e-4. This case is not exercised by the existing checkpoints.

3. **Edge-pixel fixture.** Confmap with a peak at coords `(0, 0)`, `(W-1, 0)`, `(0, H-1)`, `(W-1, H-1)`. Edge padding differs between unfold-zero-pad and `grid_sample(padding_mode="zeros")` — verify the delta is bounded and identical across runs.

4. **Cross-device determinism.** Run the new path on the same input on CPU + MPS (and CUDA when the box is available). Within a single device, identical inputs must produce bit-exact outputs across runs (seeded). Across devices, results must agree within 1e-4 — kernel implementations differ but the algorithm is the same.

5. **Single seeded run determinism.** `torch.manual_seed(0)` + identical input → identical output across two consecutive calls in the same process. (Catches accidental non-determinism from `grid_sample` falling back to a non-deterministic backend.)

## Acceptance criteria

- [ ] `ops/peaks.py::integral_refine` rewritten to use `F.grid_sample`
- [ ] `ops/crops.py::crop_bboxes` rewritten similarly
- [ ] `benchmarks/01_peak_finding.py` re-run; results appended to `benchmarks/RESULTS.md`; `find_local_peaks(integral)` shows ≥ 2× speedup at B=16
- [ ] **Per-checkpoint golden snapshot tests** for all 6 model types, generated from the pre-PR code and checked in under `tests/inference/ops/data/peak_finding_golden/`. New path matches within 1e-4 abs / 1e-5 rel.
- [ ] **Close-peak stress fixture** test (multiple peaks within `2 * integral_patch_size`) passes within 1e-4.
- [ ] **Edge-pixel fixture** test (peaks at the four image corners) passes within 1e-4 with documented expected delta.
- [ ] **Cross-device parity test** — same input on CPU + MPS + CUDA agrees within 1e-4 (skipped per-device when device unavailable; CUDA is required for merge once the CUDA box is available, otherwise gated).
- [ ] **Intra-device determinism test** — `torch.manual_seed(0)` + two consecutive calls produce identical output (`torch.equal`).
- [ ] Bottom-up end-to-end inference still matches golden output within 1e-4
- [ ] `ruff check` + `black --check` clean

## Files touched

**Modified:**
- `sleap_nn/inference/ops/peaks.py` — rewrite `integral_refine` path
- `sleap_nn/inference/ops/crops.py` — rewrite `crop_bboxes`
- `benchmarks/01_peak_finding.py` — add before/after toggle so we can compare
- `benchmarks/RESULTS.md` — append measured delta

## Out of scope

- Other GPU optimizations (torch.compile, FP16) — #511 already has opt-in flags
- Channels-last layout — explicitly deferred per §5.3
- Vectorizing the bottom-up PAF grouping (Hungarian step) — that's a CPU problem, addressed separately by #517

**Comment by gitttt-1234** [comment]
Closed — merged as part of the inference refactor stack (epic #508).
